internal/partinfo/partinfo.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package partinfo
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+)
+
+type FSType string
+
+const (
+	FAT32 FSType = "fat32"
+	Ext   FSType = "ext"
+)
+
+type PartInfo interface {
+	GetDevNode() (string, error)
+	GetLabel() (string, error)
+	GetFSType() (FSType, error)
+}
+
+type partition struct {
+	path   string
+	label  string
+	fsType FSType
+}
+
+func (b *partition) parseSuperblock() error {
+	f, err := os.OpenFile(b.path, os.O_RDONLY, 0)
+	if err != nil {
+		return fmt.Errorf("cannot parse superblock: %w", err)
+	}
+	defer f.Close()
+
+	// Attempt to read first sector of the partition.
+	s := make([]byte, 512)
+	n, err := f.Read(s)
+	if err != nil {
+		return fmt.Errorf("cannot parse superblock: %w", err)
+	}
+	if n != len(s) {
+		return fmt.Errorf("cannot parse superblock: cannot read first sector")
+	}
+
+	// Attempt to parse FAT32 superblock
+	if sb, err := newFat32Superblock(bytes.NewReader(s)); err == nil {
+		b.label = strings.TrimSpace(string(sb.Label[:]))
```

Comment:
It's mildly annoying the caller has to do this trimming instead of the specific function. I wonder if it'd be better if `newFat32Superblock` internally parsed into a struct like the one you have, but then returned a "higher level" struct with only the fields you need, and the label as a Go string. Not sure.